Clarify which keys are present in ECT authority field (#8)
ietf-rats-wg/draft-smith-rats-evidence-trans: draft-smith-rats-evidence-trans.md

> > If _m_.`notTcb` = 1 AND _f_.`notTcb` = 0; **set**(ECT.`element-list`.`element-map`.`measurement-values-map`.`flags`.`is-tcb` = TRUE).

{: dtt-enum}
* The ECT.`authority` field is set up based on the signer of the certificate containing DTI as described in {{sec-authority}}.

The completed ECT is added to the `ae` list.

## Authority field in DICE/SPDM ECTs {#sec-authority}

The ECT authority field is an array of `$crypto-keys-type-choice`s.

> When adding Evidence to the ACS, the Verifier SHALL add the public key representing the signer of the element (for example the DICE certificate or SPDM MEASUREMENTS response) to this array.

Author: ```suggestion
When adding Evidence to the ACS, the Verifier SHALL add the public key representing the signer of the evidence (for example the DICE certificate or SPDM MEASUREMENTS response) to the ECT authority field.
```

> When adding Evidence to the ACS, the Verifier SHALL add the public key representing the signer of the element (for example the DICE certificate or SPDM MEASUREMENTS response) to this array.
> The Verifier SHALL also add the signer of each certificate which has authorized the signer of the element.

Author: ```suggestion
When adding Evidence to the ACS, the Verifier SHALL add the public key representing the signer of that Evidence (for example the DICE certificate or SPDM MEASUREMENTS response) to the ECT authority field.
The Verifier SHALL also add the signer of each certificate which has authorized the signer of the signing key.
```

> The Verifier SHALL also add the signer of each certificate which has authorized (either directly or indirectly) the signer of the element.

Reviewer: Not sure what examples of directly or indirectly are.

Author: I am trying to distinguish between the signer of SPDM evidence, or the signer of a certificate which authorized a key which signed the SPDM evidence. I am sure this could be tightened up.

> The Verifier SHALL also add the signer of each certificate which has authorized (either directly or indirectly) the signer of the element.

Author: ```suggestion
The Verifier SHALL also add the signer of each certificate which has authorized the signer of the element.
```

> This lets conditional endorsers use a global key when creating endorsement conditions intended to match against evidence from multiple attestors of the same class.

Reviewer 2: ```suggestion
Having each authority in a certificate path in the ECT `authority` field lets conditional endorsement conditions  match multiple authorities within the same certificate path or match an authority that is scoped more broadly than the immediate signer of the Evidence artifact.
```

Reviewer: certificate path -> certification path (or the deprecated term certificate chain)

Author: I don't think there is a restriction on the multiple authorities being in the same certificate path. A SPDM responder may return multiple certificate chains to multiple authorities, and it would also be ok for a conditional endorser to match against the root of multiple chains.
```suggestion
Having each authority in a certificate path in the ECT `authority` field lets conditional endorsement conditions match multiple authorities or match an authority that is scoped more broadly than the immediate signer of the Evidence artifact.
```

> Each signer identity MUST be expressed using the COSE_Key varient of `$crypto-key-type-choice`.

Author: [https://ietf-rats-wg.github.io/draft-ietf-rats-corim/draft-ietf-rats-corim.html#section-8.5.3](url) contains this text:

> When comparing two $crypto-key-type-choice fields for equality, a Verifier SHALL treat them as equal if their deterministic CBOR encoding is binary equal.

When we discussed that text, there was a general desire to simplify the Verifier matching algorithm by saying that they don't need to do format comparisons.

I see the following options for making life easy for CoRIM authors:
1. Prefer a particular format when inserting data into the evidence. Verifier will need to transform to that format during evidence import.
2. Hope that all Verifiers will pick the same representation during evidence import and have the CoRIM author pick the format which their Verifier used
3. Change the CoRIM spec to let verifiers perform certain supported transforms when matching authorities
I chose option 1 but I think the other options could be made to work.

With option 1 as currently written, I don't think the CoRIM spec needs to mandate any transforms. Option 2 essentially requires transforms as otherwise it won't work when the evidence import in the verifier guesses incorrectly.

I really don't like 2. I don't really have a preference for 1 or 3 - both require that the verifier does transforms, but at different points in the verification. I think 1 is simpler.

> Each signer identity MUST be expressed using the COSE_Key varient of `$crypto-key-type-choice`.

Reviewer 2: > In general, COSE_Key is a good choice, but as Ned said, the actual internal representation is nor prescribed and CoRIM does not prescribe COSE_Key in its current form.

Unlike cases where `crypto-key-type-choice` is exposed as part of an external representation, the cert path keys are in whatever representation was used when certs were minted (e.g., ber PKCS1). Since there isn't a crypto-key-type-choice for PKCS1 the verifier has to convert it to something or we should add PKCS1 to the list of possible formats. CoRIM policy is to let the author of the external representation pick. But that doesn't work here. However, if we pick COSE_Key there will be implementations using PQC that can't accommodate large key structures. Maybe they will want to use a keyid/fingerprint approach.  I don't think there is universal enough adoption of COSE that all implementations will agree to it. The CoRIM internal representation is only a suggestion, so justifying COSE based on the assumption that the intrep is CBOR won't hold up. My choice is option 4, which is add pkcs1 as a crypto-key-type-choice, and then document using pkcs1 (which is just a copy operation). It is still possible for implementations to do key translation as a way to improve robustness. But that seems pragmatically out of scope for CoRIM / evidence translation drafts at this point.

> Each signer identity MUST be expressed using the COSE_Key varient of `$crypto-key-type-choice`.

Author: I think your option 4 is similar to option 2, but with a slightly different format.

I have changed this line to remove the preference, and put that into an issue.

```suggestion
Each signer identity MUST be expressed using `$crypto-key-type-choice`.
```

> Each signer authority MUST be expressed using the COSE_Key varient of `$crypto-key-type-choice`.

Reviewer 2: ```suggestion
Each signer authority value MUST be represented in a format that can be translated to `tagged-cose-key-type`.
```
This wording allows for CoRIM to document key translation requirements while not overstepping into CoRIM normative.

> Each signer authority MUST be expressed using the COSE_Key varient of `$crypto-key-type-choice`.

Author: ```suggestion
Each signer authority MUST be expressed using `$crypto-key-type-choice`.
```

The TCG DICE Concise Evidence Binding for SPDM specification {{-ce}} describes a process for converting the SPDM Measurement Block to Concise Evidence.
Subsequently the transformation steps defined in {{sec-ce-trans}}.

> The keys provided in the ECT.`authority` field should include global (not device-specific) keys which signed the SPDM MEASUREMENTS response carrying the Evidence as described in {{sec-authority}}.

Reviewer 2: I don't think this text is clear. All keys are global if they are secure. A key that exists on a device is device specific. If this simply references the sec-authority section wouldn't that be sufficient?

Reviewer: what was to problem to be prevented here by excluding "device-specific keys"?

Author: What I mean by a global key is one which is not device-specific.

The reason I suggest omitting device-specific keys is because they aren't that useful for authority matches. DICE keys below the DeviceID key are random and device specific, so a CoRIM author can't match against them unless they have already seen that device with those measurements.

So not really a problem, but reduces work.

```suggestion
The keys provided in the ECT.`authority` field SHOULD include the DeviceID key which authorized the SPDM MEASUREMENTS response carrying the Evidence as described in {{sec-authority}}, and keys which authorized the DeviceID key.
```

Author: For DICE/SPDM I don't think the Alias key (which signs the SPDM evidence) is very useful, but it is harmless to recommend that it be added.

```suggestion
The keys provided in the ECT.`authority` field SHOULD include the key which signed the SPDM MEASUREMENTS response carrying the Evidence as described in {{sec-authority}}, the DeviceID key which authorized that key and keys which authorized the DeviceID key.```

> The keys provided in the ECT.`authority` field SHOULD include the key which signed the SPDM MEASUREMENTS response carrying the Evidence as described in {{sec-authority}}, the DeviceID key which authorized that key and keys which authorized the DeviceID key.```

Author: ```suggestion
The keys provided in the ECT.`authority` field SHOULD include the key which signed the SPDM MEASUREMENTS response carrying the Evidence and keys which authorized that key as described in {{sec-authority}}.```
```
